v2/cdc-parent/cdc-change-applier/src/main/java/com/google/cloud/dataflow/cdc/applier/MergeStatementBuildingFn.java

```diff
@@ -1,241 +1,241 @@
+
+    // Start by actually fetching whether we created the table or not.
+    if (!createdCache) {
+      Boolean actuallyCreated = firstNonNull(tableCreated.read(), false);
+      createdCache = actuallyCreated;
+    }
+    // Once we know for sure if we created the table, then we act on creating (or not).
+    if (!createdCache) {
+      tableCreated.write(true);
+      createdCache = true;
+      c.output(KV.of(tableAndSchemas.getKey(),
+          buildCreateTableAction(tableAndSchemas, projectId, replicaDatasetId)));
+      this.tablesCreated.inc();
+    }
+
+    c.output(KV.of(tableAndSchemas.getKey(),
+        buildMergeStatementAction(
+            tableAndSchemas, projectId, changelogDatasetId, replicaDatasetId)));
+    this.mergeStatementsIssued.inc();
+  }
+
+  static final BigQueryAction buildCreateTableAction(
+      KV<String, KV<Schema, Schema>> tableAndSchemas,
+      String projectId,
+      String replicaDatasetId) {
+    Schema tableSchema = tableAndSchemas.getValue().getValue();
+    String tableName = ChangelogTableDynamicDestinations.getBigQueryTableName(
+        tableAndSchemas.getKey(), false);
+    return BigQueryAction.createTable(
+        projectId, replicaDatasetId, tableName, tableSchema);
+  }
+
+  static final BigQueryAction buildMergeStatementAction(
+      KV<String, KV<Schema, Schema>> tableAndSchemas,
+      String projectId,
+      String changelogDatasetId,
+      String replicaDatasetId) {
+    String tableName = ChangelogTableDynamicDestinations.getBigQueryTableName(
+        tableAndSchemas.getKey(), false);
+
+    String changeLogTableName = ChangelogTableDynamicDestinations.getBigQueryTableName(
+        tableAndSchemas.getKey(), true);
+
+    Schema primaryKeySchema = tableAndSchemas.getValue().getKey();
+    List<String> pkColumnNames = primaryKeySchema.getFields().stream()
+        .map(f -> f.getName()).collect(Collectors.toList());
+
+    Schema allColumnsSchema = tableAndSchemas.getValue().getValue();
+    List<String> allColumnNames = allColumnsSchema.getFields().stream()
+        .map(f -> f.getName()).collect(Collectors.toList());
+
+    String mergeStatement = buildQueryMergeReplicaTableWithChangeLogTable(
+        tableName, changeLogTableName,
+        pkColumnNames, allColumnNames,
+        projectId, changelogDatasetId, replicaDatasetId);
+    LOG.debug("Merge statement for table {} is: {}", tableName, mergeStatement);
+    return BigQueryAction.query(mergeStatement);
+  }
+
+  static final String CREATE_TABLE_TEMPLATE = String.join("",
+      "CREATE TABLE IF NOT EXISTS `%s.%s.%s` ");
+
+  static final String MERGE_REPLICA_WITH_CHANGELOG_TABLES = String.join("",
+      "MERGE `%s.%s.%s` AS replica ",
+      "USING (%s) AS changelog ",
+      "ON %s ",
+      "WHEN MATCHED AND changelog.operation = \"DELETE\" ",
+      "THEN DELETE ",
+      "WHEN MATCHED AND changelog.operation = \"UPDATE\" THEN %s ",
+      "WHEN NOT MATCHED BY TARGET AND changelog.operation = \"INSERT\" THEN %s");

```

Comment:
In this case, we can have a INSERT then UPDATE of the same primary key before a merge occurs. When this happens, we'll have no match on the target, but `operation == INSERT => FALSE` - and the action we actually want to do is to insert.